File: tools/go-by-example/runtime-config.rb
# frozen_string_literal: true
# Sprint: #118; Story: #3; Story-ID: fa07603b71dc
require_relative '../corpus/executor'

module GoByExampleRuntimeConfig
  module_function

  def configure(go, root, env, deadline:, log_prefix:)
    stages = []
    [[go, 'telemetry', 'off'], [go, 'env', '-json', 'GOTELEMETRY', 'GOTELEMETRYDIR']].each_with_index do |argv, index|
      budget = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
      raise Corpus::ContractError, 'runtime configuration deadline expired' unless budget.positive?
      stage = Corpus.capture(argv, cwd: root, env: env, timeout: [budget, 20].min,
                             stdin: File::NULL, log_prefix: log_prefix + "/#{index}")
      stages << stage
      raise Corpus::ContractError, 'SDK telemetry configuration failed' unless Corpus.success?(stage)
    end
    config = JSON.parse(File.binread(stages.last.fetch('stdout').fetch('path')))
    raise Corpus::ContractError, 'Go telemetry did not report off' unless config['GOTELEMETRY'] == 'off'
    mode = File.join(config.fetch('GOTELEMETRYDIR'), 'mode')
    raise Corpus::ContractError, 'telemetry configuration escaped isolated HOME' unless File.realpath(mode).start_with?(File.realpath(root) + '/')
    {'state' => 'complete', 'environment' => {'OTEL_TRACES_EXPORTER' => env.fetch('OTEL_TRACES_EXPORTER')},
     'go_mode' => 'off', 'mode_file' => Corpus.file_record(mode), 'stages' => stages}
  rescue StandardError => error
    {'state' => 'configuration_failure', 'detail' => error.message, 'stages' => stages}
  end
end

File: tools/go-by-example/gbe/runtimeconfig.go
// Sprint: #155; Story: S155.10; Story-ID: 67bdd9fae2b3
//
// Supported telemetry opt-outs, configured identically in every mode's
// isolated HOME before the effect baseline is taken. Ported from
// tools/go-by-example/runtime-config.rb.
package main

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var reTelemetryMode = regexp.MustCompile(`\Aoff(?: \d{4}-\d{2}-\d{2})?\z`)

// configureRuntime is GoByExampleRuntimeConfig.configure: `go telemetry off`
// then `go env -json GOTELEMETRY GOTELEMETRYDIR`, both captured, then the mode
// file proved to live inside the isolated HOME.
func configureRuntime(goBinary string, root string, env map[string]string, deadline float64, logPrefix string) *Object {
	stages := []any{}
	fail := func(message string) *Object {
		return Obj("state", "configuration_failure", "detail", message, "stages", stages)
	}
	commands := [][]string{{goBinary, "telemetry", "off"}, {goBinary, "env", "-json", "GOTELEMETRY", "GOTELEMETRYDIR"}}
	var last *Object
	for index, argv := range commands {
		budget := deadline - monotonicSeconds()
		if !(budget > 0) {
			return fail("runtime configuration deadline expired")
		}
		// `[budget, 20].min`: the Integer 20 unless the Float budget is smaller.
		timeout := Int(20)
		if budget < 20 {
			timeout = Flt(budget)
		}
		stage, err := capture(argv, root, logPrefix+"/"+itoa(index), env, timeout, os.DevNull)
		if err != nil {
			return fail(err.Error())
		}
		stages = append(stages, stage)
		last = stage
		if !success(stage) {
			return fail("SDK telemetry configuration failed")
		}
	}
	data, err := os.ReadFile(last.Obj("stdout").Str("path"))
	if err != nil {
		return fail(err.Error())
	}
	config, err := ParseObject(data)
	if err != nil {
		return fail(err.Error())
	}
	if config.Str("GOTELEMETRY") != "off" {
		return fail("Go telemetry did not report off")
	}
	dir, ok := config.Get("GOTELEMETRYDIR").(string)
	if !ok {
		return fail("key not found: \"GOTELEMETRYDIR\"")
	}
	mode := filepath.Join(dir, "mode")
	modeReal, err := realPath(mode)
	if err != nil {
		return fail(err.Error())
	}
	rootReal, err := realPath(root)
	if err != nil {
		return fail(err.Error())
	}
	if !strings.HasPrefix(modeReal, rootReal+"/") {
		return fail("telemetry configuration escaped isolated HOME")
	}
	record, err := fileRecord(mode)
	if err != nil {
		return fail(err.Error())
	}
	return Obj("state", "complete", "environment", Obj("OTEL_TRACES_EXPORTER", env["OTEL_TRACES_EXPORTER"]),
		"go_mode", "off", "mode_file", record, "stages", stages)
}
